app/custom-oauth/server/custom_oauth_server.js

```diff
@@ -310,6 +306,9 @@ export class CustomOAuth {
 				if (!user) {
 					return;
 				}
+				if (this.mergeRoles) {
```

Comment:
moved it here

Reply:
👍  just a note it will only happen on subsequent logins.  This portion is only ran on login after the user has been created.

Reply:
mmmh, then we should move it elsewhere, because it should update always

Reply:
but the name of the hook "BeforeUpdateOrCreateUserFromExternalService" suggest another effect, or?

Reply:
Unfortunately will have to divide it up.  This location for updating users.

Then inside of `validateNewUser` by doing something like:

```
users.roles = ['role1', 'role2'];
```

Reply:
You _*might*_ be able to some how add a field on serviceData an then else where manage to pull the roles off of the object and then run logic to insert it.  But it will be extremely dirty

Reply:
ok, so i should have direct setting of roles in validateNewUser
and the updating go roles with calls to `import { addUserRoles, removeUserFromRoles } from '../../authorization'` within BeforeUpdateOrCreateUserFromExternalService, right?

Reply:
if it were me something like this inside of validateNewUser

```
if (this.mergeRoles && user.services[this.name]) {
  const identity = user.services[this.name];

  if (Array.isArray(identity[this.rolesClaim])) {
    user.roles = identity[this.rolesClaim].filter((val) => val != 'offline_access' && val != 'uma_authorization');
  }
}
```